### DISCUSSION TITLE
External MEC-application

### DISCUSSION THREAD

**Discussion by itnor4** [InterDigitalInc/AdvantEDGE#433] [Q&A]
Hi, 

I have question regarding the setup and configuration of an external mec-app and ue-app in AdvantEDGE, as instructed here at https://interdigitalinc.github.io/AdvantEDGE/docs/overview/features/ext-nodes/ I have followed these configurations:

Egress mapping for mec-app     my-service::131.114.55.221:8080:TCP
Ingress mapping for ue-app       31000:my-service:8080:TCP

Where as the mec-app is a VLC application running in docker container managed by K8s and exposed at 131.114.55.221:8080
While, the ue-app is VLC application running on Ubuntu OS. 

1. ue-app receives the video at 131.114.55.221:31000 with high quality, however the throughput is not displayed correctly is lower than the actual value. For ex. the value shown is around 0.02 but the actual is around 2-3Mbps. What could be wrong?
2. When the UL/DL throughput is changed for PoA elements from 1000Mbps to 1Mbps, I still do not see worsening of the video, however the worsening of the video was observed when the packet loss is changed from low to high value. Is there any element/pod that is misconfigured for the throughput?

Thank you, I look forward to hearing from you.
 
BR,
Prachi

**Comment by itnor4** [comment]
Hi @dilallkx,

1. As referred to my above question, I have some additional information probably that could help to see where is the problem such as the log of the UE app and Wireshark traffic obtained using "kubectl sniff" on the UE application pod. 

Please find the attached scenario where as the UL, DL throughput for wifi1 is set to 1Mbps. 
![scenario](https://github.com/InterDigitalInc/AdvantEDGE/assets/94058999/52a31e3b-b88e-44a8-b73a-cb5d42a81944) and Wireshark traffic 
[wireshark.pdf](https://github.com/InterDigitalInc/AdvantEDGE/files/11664576/wireshark.pdf)                                                                       

lab2@lab2:~$ kubectl get pod --all-namespaces -o wide
NAMESPACE     NAME                                             READY   STATUS    RESTARTS      AGE     IP               NODE   NOMINATED NODE   READINESS GATES
create        mec-app-867c4b8c6c-gtpl6                         1/1     Running   0             79s     10.32.0.16       lab2   <none>           <none>
create        meep-ams-698f7fcc67-lqngr                        1/1     Running   0             19h     10.32.0.11       lab2   <none>           <none>
create        meep-app-enablement-5fcc6954c7-t59zx             1/1     Running   0             19h     10.32.0.13       lab2   <none>           <none>
create        meep-gis-engine-7945975777-8xrmh                 1/1     Running   0             19h     10.32.0.8        lab2   <none>           <none>
create        meep-loc-serv-5655c4d879-6m5kc                   1/1     Running   0             19h     10.32.0.4        lab2   <none>           <none>
create        meep-metrics-engine-78674f9656-j9gj4             1/1     Running   0             19h     10.32.0.9        lab2   <none>           <none>
create        meep-mg-manager-58fdb9f867-fqmtk                 1/1     Running   0             19h     10.32.0.5        lab2   <none>           <none>
create        meep-rnis-6b994648f8-5k5lf                       1/1     Running   0             19h     10.32.0.10       lab2   <none>           <none>
create        meep-sandbox-ctrl-88849b686-q9vmd                1/1     Running   0             19h     10.32.0.12       lab2   <none>           <none>
create        meep-tc-engine-78c77bf88d-v89b9                  1/1     Running   0             19h     10.32.0.14       lab2   <none>           <none>
create        meep-vis-85c7d99955-s776f                        1/1     Running   0             19h     10.32.0.7        lab2   <none>           <none>
create        meep-wais-565f88967c-lpxb9                       1/1     Running   0             19h     10.32.0.6        lab2   <none>           <none>
create        vlc1-56dd69f659-lzrjc                            1/1     Running   0             79s     10.32.0.35       lab2   <none>           <none>
default       mec-app                                          1/1     Running   0             2m31s   10.32.0.15       lab2   <none>           <none>
default       meep-auth-svc-6bd7b55dd8-l6ksz                   1/1     Running   0             20h     10.32.0.30       lab2   <none>           <none>
default       meep-couchdb-couchdb-0                           1/1     Running   0             20h     10.32.0.17       lab2   <none>           <none>
default       meep-docker-registry-86b6b6b849-wnvjs            1/1     Running   0             20h     10.32.0.18       lab2   <none>           <none>
default       meep-grafana-66fff596d-g89hj                     1/1     Running   0             20h     10.32.0.26       lab2   <none>           <none>
default       meep-influxdb-0                                  1/1     Running   0             20h     10.32.0.20       lab2   <none>           <none>
default       meep-ingress-controller-5s6pq                    1/1     Running   0             20h     131.114.55.221   lab2   <none>           <none>
default       meep-ingress-defaultbackend-5d8778d54f-z9gh9     1/1     Running   0             20h     10.32.0.19       lab2   <none>           <none>
default       meep-kube-state-metrics-6c75f9d88c-rgk4j         1/1     Running   0             20h     10.32.0.21       lab2   <none>           <none>
default       meep-mon-engine-7b9d647b49-5p4vv                 1/1     Running   0             20h     10.32.0.31       lab2   <none>           <none>
default       meep-open-map-tiles-8698cccbfb-mrl46             1/1     Running   0             20h     10.32.0.24       lab2   <none>           <none>
default       meep-platform-ctrl-bc6488fbc-skvhx               1/1     Running   0             20h     10.32.0.32       lab2   <none>           <none>
default       meep-postgis-0                                   2/2     Running   0             20h     10.32.0.23       lab2   <none>           <none>
default       meep-prometheus-operator-7d979b99b-zstwc         1/1     Running   0             20h     10.32.0.22       lab2   <none>           <none>
default       meep-prometheus-prometheus-node-exporter-qg9jh   1/1     Running   0             20h     131.114.55.221   lab2   <none>           <none>
default       meep-redis-master-0                              2/2     Running   0             20h     10.32.0.29       lab2   <none>           <none>
default       meep-redis-slave-0                               2/2     Running   0             20h     10.32.0.28       lab2   <none>           <none>
default       meep-virt-engine-7b4cbbcd7f-tz2ff                1/1     Running   0             20h     10.32.0.34       lab2   <none>           <none>
default       meep-webhook-f97d4599b-hqh5g                     1/1     Running   0             20h     10.32.0.33       lab2   <none>           <none>
default       prometheus-meep-prometheus-prometheus-0          2/2     Running   0             20h     10.32.0.27       lab2   <none>           <none>
kube-system   coredns-787d4945fb-9tnxw                         1/1     Running   2 (21h ago)   14d     10.32.0.3        lab2   <none>           <none>
kube-system   coredns-787d4945fb-xrwgn                         1/1     Running   2 (21h ago)   14d     10.32.0.2        lab2   <none>           <none>
kube-system   etcd-lab2                                        1/1     Running   2 (21h ago)   14d     131.114.55.221   lab2   <none>           <none>
kube-system   kube-apiserver-lab2                              1/1     Running   4 (21h ago)   14d     131.114.55.221   lab2   <none>           <none>
kube-system   kube-controller-manager-lab2                     1/1     Running   5 (21h ago)   14d     131.114.55.221   lab2   <none>           <none>
kube-system   kube-proxy-7zf24                                 1/1     Running   2 (21h ago)   14d     131.114.55.221   lab2   <none>           <none>
kube-system   kube-scheduler-lab2                              1/1     Running   5 (21h ago)   14d     131.114.55.221   lab2   <none>           <none>
kube-system   weave-net-f2bhr                                  2/2     Running   5 (21h ago)   14d     131.114.55.221   lab2   <none>           <none>


lab2@lab2:~$ kubectl logs vlc1-56dd69f659-lzrjc --namespace create
Defaulted container "meep-tc-sidecar" out of: meep-tc-sidecar, init-kube-dns (init)
time="2023-06-06T11:56:50.076Z" level=info msg="[/meep-tc-sidecar]" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.076Z" level=info msg="Starting TC Sidecar" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.076Z" level=info msg="MEEP_POD_NAME: vlc1" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.076Z" level=info msg="MEEP_SANDBOX_NAME: create" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.076Z" level=info msg="MEEP_SCENARIO_NAME: create-scenario1" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.076Z" level=info msg="Creating new MsgQueue" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.076Z" level=debug msg="Redis Connector connecting to meep-redis-master.default.svc.cluster.local:6379" meep.component=meep-tc-sidecar meep.from="db.go:69"
time="2023-06-06T11:56:50.077Z" level=info msg="Redis Connector connected to meep-redis-master.default.svc.cluster.local:6379" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.077Z" level=info msg="Successfully connected to DB" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.077Z" level=info msg="Connected to Message Queue Redis DB" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.077Z" level=info msg="Message Queue created" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.077Z" level=info msg="Successfully created new IPTables client" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.077Z" level=debug msg="Redis Connector connecting to meep-redis-master.default.svc.cluster.local:6379" meep.component=meep-tc-sidecar meep.from="db.go:69"
time="2023-06-06T11:56:50.078Z" level=info msg="Redis Connector connected to meep-redis-master.default.svc.cluster.local:6379" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.078Z" level=info msg="Successfully connected to DB" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.078Z" level=info msg="Connected to redis DB" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.078Z" level=debug msg="Redis Connector connecting to meep-redis-master.default.svc.cluster.local:6379" meep.component=meep-tc-sidecar meep.from="db.go:69"
time="2023-06-06T11:56:50.078Z" level=info msg="Redis Connector connected to meep-redis-master.default.svc.cluster.local:6379" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.078Z" level=info msg="Successfully connected to DB" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.078Z" level=info msg="Connected to Metrics Redis DB" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.078Z" level=debug msg="InfluxDB Connector connecting to http://meep-influxdb.default.svc.cluster.local:8086" meep.component=meep-tc-sidecar meep.from="metric-store.go:119"
time="2023-06-06T11:56:50.079Z" level=info msg="InfluxDB Connector connected to http://meep-influxdb.default.svc.cluster.local:8086 version: 1.8.0" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.079Z" level=info msg="Connected to Metrics Influx DB" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.08Z" level=info msg="Store name set to: create_create_scenario1" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.08Z" level=info msg="Successfully create Metric Store" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.084Z" level=info msg="Tc log update: tc qdisc change dev ifb1 handle 1:0 root netem delay 24ms 14ms 50% distribution normal loss 19.000000% rate 6000000bit" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.084Z" level=debug msg="refreshNcRules: execution time for 4 updates, elapsed time: 2.646771ms" meep.component=meep-tc-sidecar meep.from="main.go:336"
time="2023-06-06T11:56:50.084Z" level=debug msg="Fetching nat table chains" meep.component=meep-tc-sidecar meep.from="main.go:346"
time="2023-06-06T11:56:50.086Z" level=debug msg="Creating MEEP chain MEEP-ME-SERVICES" meep.component=meep-tc-sidecar meep.from="main.go:372"
time="2023-06-06T11:56:50.086Z" level=debug msg="Creating MEEP chain MEEP-INGRESS-SERVICES" meep.component=meep-tc-sidecar meep.from="main.go:384"
time="2023-06-06T11:56:50.087Z" level=debug msg="Creating MEEP chain MEEP-EGRESS-SERVICES" meep.component=meep-tc-sidecar meep.from="main.go:396"
time="2023-06-06T11:56:50.09Z" level=debug msg="Missing LB Pod IP address for pod: mec-app" meep.component=meep-tc-sidecar meep.from="main.go:524"
time="2023-06-06T11:56:50.09Z" level=debug msg="refreshLbRules: execution time: 5.989516ms" meep.component=meep-tc-sidecar meep.from="main.go:470"
time="2023-06-06T11:56:50.091Z" level=info msg="Subscription Message: subscribe to channel mq:create. Total subscriptions: 1" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:50.426Z" level=debug msg="RX MSG: Message[TC-NET-RULES-UPDATE] Src[create:meep-tc-engine] Dst[create:meep-tc-sidecar] Scope[mq:create] Payload[map[transaction-id:4136]]" meep.component=meep-tc-sidecar meep.from="main.go:318"
time="2023-06-06T11:56:50.461Z" level=debug msg="Filter created: 1 ifb: 1" meep.component=meep-tc-sidecar meep.from="main.go:873"
time="2023-06-06T11:56:50.461Z" level=debug msg="refreshNcRules: execution time for 1 updates, elapsed time: 34.495965ms" meep.component=meep-tc-sidecar meep.from="main.go:336"
time="2023-06-06T11:56:50.461Z" level=debug msg="RX MSG: Message[TC-LB-RULES-UPDATE] Src[create:meep-tc-engine] Dst[create:meep-tc-sidecar] Scope[mq:create] Payload[map[]]" meep.component=meep-tc-sidecar meep.from="main.go:315"
time="2023-06-06T11:56:50.461Z" level=debug msg="Fetching nat table chains" meep.component=meep-tc-sidecar meep.from="main.go:346"
time="2023-06-06T11:56:50.464Z" level=debug msg="Creating new service chain mapping for service: MEEP-INGRESS-SVC-MY-SERVICE-31000" meep.component=meep-tc-sidecar meep.from="main.go:553"
time="2023-06-06T11:56:50.464Z" level=debug msg="Creating MEEP chain MEEP-INGRESS-SVC-XRHEGFXC" meep.component=meep-tc-sidecar meep.from="main.go:558"
time="2023-06-06T11:56:50.468Z" level=debug msg="refreshLbRules: execution time: 7.247934ms" meep.component=meep-tc-sidecar meep.from="main.go:470"
time="2023-06-06T11:56:55.194Z" level=info msg="Pinger error: 10.32.0.16 i/o timeout" meep.component=meep-tc-sidecar
time="2023-06-06T11:56:59.197Z" level=info msg="Pinger error: 10.32.0.16 i/o timeout" meep.component=meep-tc-sidecar
time="2023-06-06T11:57:00.422Z" level=debug msg="RX MSG: Message[TC-NET-RULES-UPDATE] Src[create:meep-tc-engine] Dst[create:meep-tc-sidecar] Scope[mq:create] Payload[map[transaction-id:4137]]" meep.component=meep-tc-sidecar meep.from="main.go:318"
time="2023-06-06T11:57:00.423Z" level=debug msg="refreshNcRules: execution time for 0 updates, elapsed time: 554.809µs" meep.component=meep-tc-sidecar meep.from="main.go:336"
time="2023-06-06T11:57:00.923Z" level=debug msg="RX MSG: Message[TC-NET-RULES-UPDATE] Src[create:meep-tc-engine] Dst[create:meep-tc-sidecar] Scope[mq:create] Payload[map[transaction-id:4138]]" meep.component=meep-tc-sidecar meep.from="main.go:318"
time="2023-06-06T11:57:00.924Z" level=debug msg="refreshNcRules: execution time for 0 updates, elapsed time: 909.285µs" meep.component=meep-tc-sidecar meep.from="main.go:336"
time="2023-06-06T11:57:04.2Z" level=info msg="Pinger error: 10.32.0.16 i/o timeout" meep.component=meep-tc-sidecar
time="2023-06-06T11:57:05.2Z" level=info msg="Pinger error: 10.32.0.16 i/o timeout" meep.component=meep-tc-sidecar
time="2023-06-06T11:57:06.2Z" level=info msg="Pinger error: 10.32.0.16 i/o timeout" meep.component=meep-tc-sidecar
time="2023-06-06T11:57:08.202Z" level=info msg="Pinger error: 10.32.0.16 i/o timeout" meep.component=meep-tc-sidecar
time="2023-06-06T11:57:09.202Z" level=info msg="Pinger error: 10.32.0.16 i/o timeout" meep.component=meep-tc-sidecar
time="2023-06-06T11:57:14.207Z" level=info msg="Pinger error: 10.32.0.16 i/o timeout" meep.component=meep-tc-sidecar
time="2023-06-06T11:57:14.423Z" level=debug msg="RX MSG: Message[TC-NET-RULES-UPDATE] Src[create:meep-tc-engine] Dst[create:meep-tc-sidecar] Scope[mq:create] Payload[map[transaction-id:4139]]" meep.component=meep-tc-sidecar meep.from="main.go:318"
time="2023-06-06T11:57:14.423Z" level=debug msg="refreshNcRules: execution time for 0 updates, elapsed time: 525.442µs" meep.component=meep-tc-sidecar meep.from="main.go:336"
time="2023-06-06T11:57:14.925Z" level=debug msg="RX MSG: Message[TC-NET-RULES-UPDATE] Src[create:meep-tc-engine] Dst[create:meep-tc-sidecar] Scope[mq:create] Payload[map[transaction-id:4140]]" meep.component=meep-tc-sidecar meep.from="main.go:318"
time="2023-06-06T11:57:14.926Z" level=debug msg="refreshNcRules: execution time for 0 updates, elapsed time: 594.481µs" meep.component=meep-tc-sidecar meep.from="main.go:336"

2. What is the component in AdvantEDGE that adds the impairment to the application traffic, is it different component for packet loss and throughput?

Thank you in advance, I look forward to hearing from you.

BR,
Prachi

**Comment by dilallkx** [comment]
Hi @itnor4,

Thank you for using the platform & for reaching out!

It seems your traffic is not going through the platform as expected. I would like to have a closer look at your scenario to verify the ingress and egress rules for the external apps. Could you please export and attach your scenario to this discussion?

Could you also clarify where your external apps are running?
I assume vlc1 is the VLC client running on Ubuntu OS; is it running on the same machine as AdvantEDGE or on a different setup?
I assume mec-app is the VLC server running in a docker container; how was it deployed? Is it on the same setup as AdvantEDGE?

Also, what are the expected traffic flows that you want to pass through AdvantEDGE? For example, the VLC client may send a request to the VLC server on port 31000 via AdvantEDGE, however the subsequent data flow may not use the same session & ports, and may require a different set of rules to pass through AdvantEDGE. 

With this additional information we should be able to get this working for you!
Regards!

**Comment by itnor4** [comment]
Hi @dilallkx,

Thank you for your response and please find the attach scenario.  
[single-meh.yaml.zip](https://github.com/InterDigitalInc/AdvantEDGE/files/11708934/single-meh.yaml.zip)
Yes, it seems that the traffic does not follow the path as expected, the UE application pod created by AdvantEDGE log shows initial traffic and packets, also within the Wireshark this has been observed. This only occurs with TCP not with UDP (UDP works fine.)

Could you also clarify where your external apps are running? 
The external application is an VLC app running on Ubuntu OS in different machine with an IP address of 131.114.55.150, I am trying to receive the stream at 131.114.55.221:31000. 

The mec-app is deployed at the same machine where AdvantEDGE is running, I am using VLC docker container in a pod that plays the video on loop. I have created service for that pod mapping the external IP as 131.114.55.221 with port 8080 (This IP address is same as AdvantEDGE).

For ex. during different practical trials I have also used VLC application to stream the video with TCP instead of container in K8s pod, the issue remains the same. If the packet loss is too high at PoA where the UE is connected then I see effect of the network characteristics being applied such as (packet loss effect), when the packet loss is lower than 40% around 20-30% there is no effect the video still runs with good quality and the UE app pod log always has this issue (time="2023-06-06T11:57:04.2Z" level=info msg="Pinger error: 10.32.0.16 i/o timeout" meep.component=meep-tc-sidecar) running in the backend. The throughput is calculated wrong. 

However, the UDP did not had any such problems it works well as expected. 

Thank you, I look forward to hearing from you.

BR,
Prachi

**Comment by dilallkx** [comment]
Hi @itnor4,

Thank you for the scenario.

If I understand correctly, you have a VLC app running in a k8s pod on the same machine as AdvantEDGE that sends a TCP video stream to the VLC player running on a different Ubuntu machine. The VLC app is configured to send the video stream to AdvantEDGE at 131.114.55.221:31000 using the mec-app ingress rule, and you would then expect it to be sent to the external VLC player at 131.114.55.150:8080 using the vlc1 egress rule. This seems correct to me.

Do you confirm that VLC protocol used with TCP creates a single TCP session between your external applications? Is it possible that multiple TCP sessions are established by the applications when using TCP? This could lead to some flows passing through AdvantEDGE and other flows that bypass AdvantEDGE. Please confirm (with traces on the external player) that there is a single TCP connection on port 8080, and that the video stream data flow uses this port. Also, please verify that the TCP SYN packet for the TCP session originates from the container and is sent to port 31000.

Regards!

**Comment by itnor4** [comment]
Hi @dilallkx,

Thank you for your response. 

Just to confirm, does the configuration of 131.114.55.150:8080 needs to be done at UE app within AdvantEDGE using Egress rule, currently the setting is at 131.114.55.221:8080 which is the AdvantEDGE and mec-app IP. 

Egress mapping for ue-app my-service::131.114.55.221:8080:TCP
Ingress mapping for mec-app 31000:my-service:8080:TCP

**Comment by dilallkx** [comment]
Hi @itnor4,

I seem to have mixed up the IP addresses in my previous reply.

The ingress rule is the AdvantEDGE **entry** point where traffic (i.e. video stream) should be sent; in your case this is mec-app 131.114.55.221:31000.
The egress rule is the AdvantEDGE **exit** point where **shaped** traffic will be forwarded to; in your case this is vlc1 131.114.55.221:8080.

With this configuration, the only traffic that will pass through AdvantEDGE is TCP traffic for the TCP session established from an external app towards 131.114.55.221:31000.

If your external VLC app establishes a TCP connection with 131.114.55.221:31000 and **sends the video stream over this same TCP connection**, then it will go through AdvantEDGE and reach the VLC player at 131.114.55.221:8080. If a new TCP session is established for the video stream on a different port or with a different protocol then it may bypass AdvantEDGE. This is where the trace is useful to identify the data flows.

I hope this clarifies the platform behavior!
Regards.
